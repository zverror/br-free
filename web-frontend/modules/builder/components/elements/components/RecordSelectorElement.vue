<template>
  <ABFormGroup
    :label="resolvedLabel"
    :required="element.required"
    :error-message="getErrorMessage()"
    :style="getStyleOverride('input')"
  >
    <ABDropdown
      ref="recordSelectorDropdown"
      v-model="inputValue"
      :show-search="adhocSearchEnabled"
      :emit-search="adhocSearchEnabled"
      class="choice-element"
      :placeholder="resolvedPlaceholder"
      :multiple="element.multiple"
      :before-show="beforeShow"
      @hide="onFormElementTouch"
      @query-change="adhocSearch = $event"
      @scroll="$refs.infiniteScroll.handleScroll($event)"
    >
      <template #value>
        <template v-if="loading">
          <div class="loading" />
        </template>
        <span class="ab-dropdown__selected-text">
          {{ selectedValueDisplay }}
        </span>
      </template>
      <template #emptyState>
        {{
          adhocSearchEnabled
            ? $t('recordSelectorElement.emptyAdhocState', {
                query: adhocSearch,
              })
            : $t('recordSelectorElement.emptyState')
        }}
      </template>
      <template #defaultValue>
        <template v-if="loading">
          <div class="loading" />
        </template>
        <span class="ab-dropdown__selected-placeholder">{{
          resolvedPlaceholder ? resolvedPlaceholder : $t('action.makeChoice')
        }}</span>
      </template>
      <InfiniteScroll
        ref="infiniteScroll"
        :current-count="currentOffset"
        :has-more-page="hasMorePage"
        :loading="loading"
        :reverse="false"
        :render-end="false"
        @load-next-page="nextPage"
      >
        <template #default>
          <ABDropdownItem
            v-for="{ value, name, nameSuffix } in resolvedOptions"
            :key="value"
            :name="`${name}${nameSuffix ? ` - ${nameSuffix}` : ''}`"
            :value="value"
          />
        </template>
      </InfiniteScroll>
    </ABDropdown>
  </ABFormGroup>
</template>

<script>
import _ from 'lodash'
import {
  ensureArray,
  ensureInteger,
  ensureString,
} from '@baserow/modules/core/utils/validator'
import formElement from '@baserow/modules/builder/mixins/formElement'
import collectionElement from '@baserow/modules/builder/mixins/collectionElement'
import RuntimeFormulaContext from '@baserow/modules/core/runtimeFormulaContext'
import InfiniteScroll from '@baserow/modules/core/components/helpers/InfiniteScroll.vue'
import { notifyIf } from '@baserow/modules/core/utils/error'
import DataSourceService from '@baserow/modules/builder/services/dataSource'

export default {
  name: 'RecordSelectorElement',
  components: { InfiniteScroll },
  mixins: [formElement, collectionElement],
  props: {
    /**
     * @type {Object}
     * @property {boolean} required - If the element is required for form submission
     * @property {string} data_source_id - The data source for the record selector element
     * @propeRty {number} items_per_page - Number of items to show per page
     * @property {string} label - The label displayed above the record selector element
     * @property {string} default_value - The formula to generate the displayed name
     * @property {string} placeholder - The placeholder text which should be applied to the element
     * @property {boolean} multiple - Whether this element can hold multiple values
     * @property {string} option_name_suffix - The formula to generate the displayed suffix name
     */
    element: {
      type: Object,
      required: true,
    },
  },
  data() {
    return {
      loading: false,
      openedOnce: false,
      defaultValueOptions: [],
      shouldDebounce: false,
      debounceTimeout: null,
    }
  },
  computed: {
    adhocSearchEnabled() {
      return (
        this.elementType.adhocSearchableProperties(
          this.element,
          this.dataSource
        ).length > 0
      )
    },
    resolvedLabel() {
      return ensureString(this.resolveFormula(this.element.label))
    },
    resolvedDefaultValue() {
      const resolvedFormula = this.resolveFormula(this.element.default_value)
      if (this.element.multiple) {
        try {
          return ensureArray(resolvedFormula).map(ensureInteger)
        } catch {
          return []
        }
      } else {
        try {
          return ensureInteger(resolvedFormula)
        } catch {
          return null
        }
      }
    },
    resolvedPlaceholder() {
      return ensureString(this.resolveFormula(this.element.placeholder))
    },
    /**
     * These fallback options are here to maintain reactivity of the dropdown component
     * that can detect DOM changes. As we have always at least fallback for missing
     * rows we can see the default value.
     */
    fallbackOptions() {
      let recordIds = this.resolvedDefaultValue

      if (!this.element.multiple) {
        if (this.resolvedDefaultValue !== null) {
          recordIds = [this.resolvedDefaultValue]
        } else {
          recordIds = []
        }
      }

      return recordIds.map((id) => ({
        value: id,
        name: this.$t('recordSelectorElementForm.record', { id }),
        nameSuffix: '',
      }))
    },
    resolvedOptions() {
      // Fill the dropdown options with an array containing
      // the record id and name from the resolved formula
      const options = this.elementContent.map((record, recordIndex) => ({
        value: record?.id,
        nameSuffix: ensureString(
          this.resolveFormula(
            this.element.option_name_suffix,
            // Formula context to resolve `current record` to the current index
            new Proxy(
              new RuntimeFormulaContext(
                this.$registry.getAll('builderDataProvider'),
                {
                  ...this.applicationContext,
                  // We don't want the full recordIndex path here as we always want a
                  // data source as source for this component and we don't want nested
                  // data selection.
                  recordIndexPath: [recordIndex],
                  allowSameElement: true,
                }
              ),
              {
                get(target, prop) {
                  return target.get(prop)
                },
              }
            )
          )
        ),
        name:
          this.dataSourceType.getRecordName(this.dataSource, record || {}) ||
          this.$t('recordSelectorElementForm.record', { id: recordIndex }),
      }))

      // Append the default value options so that they are displayed in the
      // dropdown even if we haven't reached its page yet.
      // When we retrieve the records that are present in the default values,
      // we remove the duplicate ones.
      return _.uniqBy(
        [...options, ...this.defaultValueOptions, ...this.fallbackOptions],
        'value'
      )
    },
    selectedOption() {
      if (this.element.multiple) {
        return this.resolvedOptions.filter((option) =>
          this.inputValue.includes(option.value)
        )
      } else {
        return this.resolvedOptions.find(
          (option) => option.value === this.inputValue
        )
      }
    },
    selectedValueDisplay() {
      if (this.element.multiple) {
        return this.selectedOption.map(({ name }) => name).join(', ')
      } else {
        return this.selectedOption?.name
      }
    },
    elementContent() {
      // elementContent mixin uses applicationContext.recordIndexPath to determine
      // if we are inside a collection element.
      // In this scenario we set recordIndexPath to be empty so that the whole
      // element._.content is returned
      // Also the element content can be shared among all the record selectors in all
      // repetitions, it's even better and it saves queries.
      return this.getElementContent(this.element, {
        ...this.applicationContext,
        recordIndexPath: [],
      })
    },
  },
  watch: {
    resolvedDefaultValue: {
      async handler(newDefaultValue, oldValue) {
        if (!_.isEqual(newDefaultValue, oldValue)) {
          this.inputValue = newDefaultValue
          if (!this.shouldDebounce) {
            await this.updateDefaultRecordNames(newDefaultValue)
            this.shouldDebounce = true
          } else {
            // Debounced update to avoid too many queries
            this.debouncedUpdateDefaultRecordNames(newDefaultValue)
          }
        }
      },
      immediate: true,
    },
    async 'element.data_source_id'() {
      await this.updateDefaultRecordNames(this.resolvedDefaultValue)
    },
    'element.multiple'() {
      this.setFormData(this.resolvedDefaultValue)
    },
  },
  methods: {
    async beforeShow() {
      if (!this.openedOnce) {
        this.openedOnce = true
        this.loading = true
        // Fetch first content on first opening instead of onMount.
        await this.fetchContent([0, this.element.items_per_page])
        this.loading = false
      }
    },
    async nextPage() {
      try {
        this.loading = true
        await this.loadMore()
      } catch (e) {
        notifyIf(e, 'application')
      } finally {
        this.loading = false
      }
    },
    async updateDefaultRecordNames(recordIdsToFetch) {
      // There is no need to resolve the default value when there is no data source
      if (!this.element.data_source_id) {
        this.defaultValueOptions = []
        return
      }

      let recordIds = recordIdsToFetch
      if (!this.element.multiple) {
        if (recordIdsToFetch !== null) {
          recordIds = [recordIdsToFetch]
        } else {
          recordIds = []
        }
      }

      if (recordIds.length === 0) {
        return
      }

      this.loading = true
      try {
        const service = DataSourceService(this.$client)
        const data = await service.getRecordNames(
          this.element.data_source_id,
          recordIds
        )
        this.defaultValueOptions = Object.entries(data).map(
          ([value, name]) => ({
            value: parseInt(value),
            name,
            actualName: name,
          })
        )
      } catch (e) {
        notifyIf(e, 'application')
      } finally {
        this.loading = false
      }
    },
    debouncedUpdateDefaultRecordNames(val) {
      clearTimeout(this.debounceTimeout)
      this.debounceTimeout = setTimeout(() => {
        this.updateDefaultRecordNames(val)
      }, 500)
    },
    canFetch() {
      // We want to fetch data only if the dropdown have been opened at least once.
      // It's not necessary otherwise
      return this.openedOnce && this.contentFetchEnabled
    },
    getErrorMessage() {
      return this.displayFormDataError ? this.$t('error.requiredField') : ''
    },
  },
}
</script>
