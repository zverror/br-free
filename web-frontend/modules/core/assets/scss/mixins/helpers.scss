@mixin fixed-height($height, $font-size) {
  height: $height;
  line-height: $height;
  font-size: $font-size;
}

@mixin center-text($width, $font-size, $height: $width) {
  @include fixed-height($height, $font-size);

  width: $width;
  text-align: center;
}

@mixin flex-align-items($gap: 0, $display: flex) {
  display: $display;
  align-items: center;
  gap: $gap;
}

@mixin loading($size: 1.4rem) {
  position: relative;
  display: block;
  width: $size;
  height: $size;
  border-radius: 50%;
  border: 0.25em solid;
  border-color: $color-primary-500 transparent $color-primary-500 transparent;
  animation: spin infinite 1800ms;
  animation-timing-function: cubic-bezier(0.785, 0.135, 0.15, 0.86);
}

// The basically works the same as padding: 10px 0 0 10px or padding: 10px; only then
// resulting position: absolute; with corresponding top, right, bottom, left values.
// -------
// Example:
//  @include absolute(10px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 10px; bottom: 10px; left: 10px;
// -------
// Example:
//  @include absolute(10px, 20px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 10px; left: 20px;
// -------
// Example:
//  @include absolute(10px, 20px, 30px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 30px;
@mixin absolute($top: null, $right: null, $bottom: null, $left: null) {
  position: absolute;

  @if $top != null and $right == null and $bottom == null and $left == null {
    inset: $top $top $top $top;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else if $top != null and $right != null and $bottom == null and $left == null
  {
    inset: $top $right $top $right;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else {
    @if $top {
      top: $top;
    }

    @if $right {
      right: $right;
    }

    @if $bottom {
      bottom: $bottom;
    }

    @if $left {
      left: $left;
    }
  }
}
