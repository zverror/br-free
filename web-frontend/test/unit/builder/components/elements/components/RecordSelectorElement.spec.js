import { TestApp } from '@baserow/test/helpers/testApp'
import RecordSelectorElement from '@baserow/modules/builder/components/elements/components/RecordSelectorElement.vue'
import flushPromises from 'flush-promises'

// Ignore `notifyIf` and `notifyIf404` function calls
jest.mock('@baserow/modules/core/utils/error.js')

describe('RecordSelectorElement', () => {
  let testApp = null
  let store = null
  let mockServer = null

  beforeAll(() => {
    // NOTE: TestApp wraps any exception raised by the axios mock adapter and
    // re-raises it as a Jest error.
    // This mutates the error object and make some properties not available.
    // In this case `collectionElement` mixin needs to access the response
    // object when the server returns a 400/404 error, so we disable
    // `failTestOnErrorResponse`.
    testApp = new TestApp()
    testApp.failTestOnErrorResponse = false
    store = testApp.store
    mockServer = testApp.mockServer
  })

  afterEach(() => {
    testApp.afterEach()
  })

  const mountComponent = ({ props = {}, slots = {}, provide = {} }) => {
    return testApp.mount(RecordSelectorElement, {
      propsData: props,
      slots,
      provide,
    })
  }

  test('does not paginate if API returns 400/404', async () => {
    const page = {
      id: 1,
      dataSources: [{ id: 1, type: 'local_baserow_list_rows', table_id: 1 }],
      elements: [],
    }
    const sharedPage = {
      id: 2,
      dataSources: [],
      elements: [],
      shared: true,
    }
    const builder = {
      id: 1,
      theme: { primary_color: '#ccc' },
      pages: [sharedPage, page],
    }
    const workspace = {}
    const mode = 'public'
    const element = {
      id: 1,
      type: 'record_selector',
      data_source_id: page.dataSources[0].id,
      items_per_page: 5,
    }
    store.dispatch('element/forceCreate', { page, element })

    const wrapper = await mountComponent({
      props: {
        element,
      },
      provide: {
        builder,
        currentPage: page,
        elementPage: page,
        mode,
        applicationContext: { builder, page, mode },
        element,
        workspace,
      },
    })

    // A mock server that mimics the data source dispatch endpoints.
    // The first time it is called it returns a successful message, but the
    // second time it returns a 400 response.
    const url = `builder/domains/published/data-source/${page.dataSources[0].id}/dispatch/`
    mockServer.mock
      .onPost(url)
      .replyOnce(200, {
        results: [
          { id: 1, order: 1, field_1: 'First' },
          { id: 2, order: 1, field_1: 'Second' },
          { id: 3, order: 1, field_1: 'Third' },
          { id: 4, order: 1, field_1: 'Fourth' },
          { id: 5, order: 1, field_1: 'Fifth' },
        ],
        has_next_page: true,
      })
      .onPost(url)
      .reply(400, { message: 'Bad Request' })

    // The first time we trigger a next page, the server responds with 200
    // therefore we should be able to fetch more content
    await wrapper
      .findAllComponents({ name: 'ABDropdown' })
      .at(0)
      .find('.ab-dropdown__selected')
      .trigger('click')
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()
    expect(mockServer.mock.history.post.length).toBe(1)

    // Then we trigger a few scroll events in the record selector element and
    // confirm that the API is only called once
    await wrapper
      .findAllComponents({ name: 'ABDropdown' })
      .at(0)
      .find('.select__items')
      .trigger('scroll')
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()
    expect(mockServer.mock.history.post.length).toBe(2)

    await wrapper
      .findAllComponents({ name: 'ABDropdown' })
      .at(0)
      .find('.select__items')
      .trigger('scroll')
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()
    expect(mockServer.mock.history.post.length).toBe(2)
  })

  test('resolves suffix formulas', async () => {
    const page = {
      id: 1,
      dataSources: [{ id: 1, type: 'local_baserow_list_rows', table_id: 1 }],
      elements: [],
    }
    const sharedPage = {
      id: 2,
      dataSources: [],
      elements: [],
      shared: true,
    }
    const builder = {
      id: 1,
      theme: { primary_color: '#ccc' },
      pages: [sharedPage, page],
    }
    const workspace = {}
    const mode = 'public'
    const element = {
      id: 1,
      type: 'record_selector',
      data_source_id: page.dataSources[0].id,
      items_per_page: 5,
      option_name_suffix: "'Suffix'",
    }
    store.dispatch('element/forceCreate', { page, element })

    const wrapper = await mountComponent({
      props: {
        element,
      },
      provide: {
        builder,
        currentPage: page,
        elementPage: page,
        mode,
        applicationContext: { builder, page, mode, element },
        element,
        workspace,
      },
    })

    // A simple mock server that return ids and some fields
    const url = `builder/domains/published/data-source/${page.dataSources[0].id}/dispatch/`
    mockServer.mock.onPost(url).reply(200, {
      results: [
        { id: 1, order: 1, field_1: 'First', field_2: 'One' },
        { id: 2, order: 1, field_1: 'Second', field_2: 'Two' },
      ],
      has_next_page: false,
    })

    // Check that the literal string was added to all items in the record selector
    await wrapper
      .findAllComponents({ name: 'ABDropdown' })
      .at(0)
      .find('.ab-dropdown__selected')
      .trigger('click')
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find("span[title='First - Suffix']").exists()).toBeTruthy()
    expect(wrapper.find("span[title='Second - Suffix']").exists()).toBeTruthy()

    // Set a formula for suffix and check it was properly resolved
    store.dispatch('element/forceUpdate', {
      page,
      element,
      values: { option_name_suffix: "get('current_record.field_2')" },
    })
    await flushPromises()
    await wrapper
      .findAllComponents({ name: 'ABDropdown' })
      .at(0)
      .find('.ab-dropdown__selected')
      .trigger('click')
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find("span[title='First - One']").exists()).toBeTruthy()
    expect(wrapper.find("span[title='Second - Two']").exists()).toBeTruthy()
  })
})
