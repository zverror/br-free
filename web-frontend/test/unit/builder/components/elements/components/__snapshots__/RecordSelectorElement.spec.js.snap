// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecordSelectorElement does not paginate if API returns 400/404 1`] = `
<div
  class="control ab-form-group"
>
  <!---->
   
  <!---->
   
  <div
    class="control__wrapper"
  >
    <div
      class="control__elements"
    >
      <div
        class="flex-grow-1"
      >
        <!---->
         
        <div
          class="ab-form-group__children"
        >
          <div
            class="ab-dropdown choice-element"
            tabindex="0"
          >
            <a
              class="ab-dropdown__selected"
            >
              <!---->
               
              <span
                class="ab-dropdown__selected-placeholder"
              >
                action.makeChoice
              </span>
               
              <i
                class="ab-dropdown__toggle-icon iconoir-nav-arrow-down"
              />
            </a>
             
            <div
              class="ab-dropdown__items"
            >
              <!---->
               
              <ul
                class="select__items"
                style=""
                tabindex="-1"
              >
                   
                <section
                  class="infinite-scroll"
                >
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="First"
                        >
                          First
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Second"
                        >
                          Second
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Third"
                        >
                          Third
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Fourth"
                        >
                          Fourth
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Fifth"
                        >
                          Fifth
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                   
                  <div
                    class="infinite-scroll__loading-wrapper"
                    style=""
                  >
                    <!---->
                  </div>
                   
                  <!---->
                </section>
              </ul>
               
              <!---->
               
              <!---->
            </div>
          </div>
           
          <!---->
        </div>
      </div>
       
      <!---->
       
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`RecordSelectorElement does not paginate if API returns 400/404 2`] = `
<div
  class="control ab-form-group"
>
  <!---->
   
  <!---->
   
  <div
    class="control__wrapper"
  >
    <div
      class="control__elements"
    >
      <div
        class="flex-grow-1"
      >
        <!---->
         
        <div
          class="ab-form-group__children"
        >
          <div
            class="ab-dropdown choice-element"
            tabindex="0"
          >
            <a
              class="ab-dropdown__selected"
            >
              <!---->
               
              <span
                class="ab-dropdown__selected-placeholder"
              >
                action.makeChoice
              </span>
               
              <i
                class="ab-dropdown__toggle-icon iconoir-nav-arrow-down"
              />
            </a>
             
            <div
              class="ab-dropdown__items"
            >
              <!---->
               
              <ul
                class="select__items"
                style=""
                tabindex="-1"
              >
                   
                <section
                  class="infinite-scroll"
                >
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="First"
                        >
                          First
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Second"
                        >
                          Second
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Third"
                        >
                          Third
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Fourth"
                        >
                          Fourth
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Fifth"
                        >
                          Fifth
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                   
                  <div
                    class="infinite-scroll__loading-wrapper"
                    style=""
                  >
                    <!---->
                  </div>
                   
                  <!---->
                </section>
              </ul>
               
              <!---->
               
              <!---->
            </div>
          </div>
           
          <!---->
        </div>
      </div>
       
      <!---->
       
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`RecordSelectorElement does not paginate if API returns 400/404 3`] = `
<div
  class="control ab-form-group"
>
  <!---->
   
  <!---->
   
  <div
    class="control__wrapper"
  >
    <div
      class="control__elements"
    >
      <div
        class="flex-grow-1"
      >
        <!---->
         
        <div
          class="ab-form-group__children"
        >
          <div
            class="ab-dropdown choice-element"
            tabindex="0"
          >
            <a
              class="ab-dropdown__selected"
            >
              <!---->
               
              <span
                class="ab-dropdown__selected-placeholder"
              >
                action.makeChoice
              </span>
               
              <i
                class="ab-dropdown__toggle-icon iconoir-nav-arrow-down"
              />
            </a>
             
            <div
              class="ab-dropdown__items"
            >
              <!---->
               
              <ul
                class="select__items"
                style=""
                tabindex="-1"
              >
                   
                <section
                  class="infinite-scroll"
                >
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="First"
                        >
                          First
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Second"
                        >
                          Second
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Third"
                        >
                          Third
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Fourth"
                        >
                          Fourth
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Fifth"
                        >
                          Fifth
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                   
                  <div
                    class="infinite-scroll__loading-wrapper"
                    style=""
                  >
                    <!---->
                  </div>
                   
                  <!---->
                </section>
              </ul>
               
              <!---->
               
              <!---->
            </div>
          </div>
           
          <!---->
        </div>
      </div>
       
      <!---->
       
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`RecordSelectorElement resolves suffix formulas 1`] = `
<div
  class="control ab-form-group"
>
  <!---->
   
  <!---->
   
  <div
    class="control__wrapper"
  >
    <div
      class="control__elements"
    >
      <div
        class="flex-grow-1"
      >
        <!---->
         
        <div
          class="ab-form-group__children"
        >
          <div
            class="ab-dropdown choice-element"
            tabindex="0"
          >
            <a
              class="ab-dropdown__selected"
            >
              <!---->
               
              <span
                class="ab-dropdown__selected-placeholder"
              >
                action.makeChoice
              </span>
               
              <i
                class="ab-dropdown__toggle-icon iconoir-nav-arrow-down"
              />
            </a>
             
            <div
              class="ab-dropdown__items"
            >
              <!---->
               
              <ul
                class="select__items"
                style=""
                tabindex="-1"
              >
                   
                <section
                  class="infinite-scroll"
                >
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="First - Suffix"
                        >
                          First - Suffix
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Second - Suffix"
                        >
                          Second - Suffix
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                   
                  <div
                    class="infinite-scroll__loading-wrapper"
                    style="display: none;"
                  >
                    <!---->
                  </div>
                   
                  <!---->
                </section>
              </ul>
               
              <!---->
               
              <!---->
            </div>
          </div>
           
          <!---->
        </div>
      </div>
       
      <!---->
       
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`RecordSelectorElement resolves suffix formulas 2`] = `
<div
  class="control ab-form-group"
>
  <!---->
   
  <!---->
   
  <div
    class="control__wrapper"
  >
    <div
      class="control__elements"
    >
      <div
        class="flex-grow-1"
      >
        <!---->
         
        <div
          class="ab-form-group__children"
        >
          <div
            class="ab-dropdown choice-element"
            tabindex="0"
          >
            <a
              class="ab-dropdown__selected"
            >
              <!---->
               
              <span
                class="ab-dropdown__selected-placeholder"
              >
                action.makeChoice
              </span>
               
              <i
                class="ab-dropdown__toggle-icon iconoir-nav-arrow-down"
              />
            </a>
             
            <div
              class="ab-dropdown__items"
            >
              <!---->
               
              <ul
                class="select__items"
                style=""
                tabindex="-1"
              >
                   
                <section
                  class="infinite-scroll"
                >
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="First - One"
                        >
                          First - One
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                  <li
                    class="ab-dropdownitem__item ab-dropdownitem__item--no-options"
                  >
                    <a
                      class="ab-dropdownitem__item-link"
                    >
                      <div
                        class="ab-dropdownitem__item-name"
                      >
                        <!---->
                         
                        <span
                          class="ab-dropdownitem__item-name-text"
                          title="Second - Two"
                        >
                          Second - Two
                        </span>
                      </div>
                       
                      <!---->
                    </a>
                     
                    <i
                      class="ab-dropdownitem__item-active-icon iconoir-check"
                    />
                  </li>
                   
                  <div
                    class="infinite-scroll__loading-wrapper"
                    style="display: none;"
                  >
                    <!---->
                  </div>
                   
                  <!---->
                </section>
              </ul>
               
              <!---->
               
              <!---->
            </div>
          </div>
           
          <!---->
        </div>
      </div>
       
      <!---->
       
    </div>
     
    <!---->
  </div>
</div>
`;
